DOC: Add slicer.parameterNodeWrapper autodocs

allemangD opened the pull request:
Add `slicer.parameterNodeWrapper` automodule to the Slicer Python API documentation.

---

Currently this fails because sphinx does not have access to the extension types in the Slicer namespace. See #6875 for more details.

allemangD committed:
DOC: Add slicer.parameterNodeWrapper autodocs

jcfr commented:
The following error is reported on readthedocs:

```
WARNING: autodoc: failed to import module 'parameterNodeWrapper' from module 'slicer'; the following exception was raised:
cannot import name 'Annotated' from 'typing' (/home/docs/.pyenv/versions/3.7.9/lib/python3.7/typing.py)
```

jcfr commented:
The generation is currently failing with the following error:

```
WARNING: autodoc: failed to import module 'parameterNodeWrapper' from module 'slicer'; the following exception was raised:
Traceback (most recent call last):
  [...]
  File "/home/jcfr/Projects/Slicer/Base/Python/slicer/parameterNodeWrapper/__init__.py", line 5, in <module>
    from .guiConnectors import *
  File "/home/jcfr/Projects/Slicer/Base/Python/slicer/parameterNodeWrapper/guiConnectors.py", line 16, in <module>
    from . import parameterPack as pack
  File "/home/jcfr/Projects/Slicer/Base/Python/slicer/parameterNodeWrapper/parameterPack.py", line 10, in <module>
    from .serializers import (
  File "/home/jcfr/Projects/Slicer/Base/Python/slicer/parameterNodeWrapper/serializers.py", line 402, in <module>
    class NodeSerializer(Serializer):
  File "/home/jcfr/Projects/Slicer/Base/Python/slicer/parameterNodeWrapper/serializers.py", line 424, in NodeSerializer
    def write(self, parameterNode, name: str, value: slicer.vtkMRMLNode) -> None:
AttributeError: module 'slicer' has no attribute 'vtkMRMLNode'
```

jcfr committed:
BUG: Allow import of parameterNodeWrapper from standalone python console

This commit allows to successfully import the package from a standalone
python console (e.g PythonSlicer).

Supporting this is helpful in the context of documentation generation and
IDE integration.

jcfr committed:
BUG: Fix support for generating sphinx doc from Slicer Python environment

This commit fixes the error reported below by ensuring the path to the
"slicer" package available in the source tree is prepended only if
not already available from the Slicer Python environment where modules
like MRMLCorePython have been successfully imported.

Note that since "slicer.kits" is only available in the Slicer Python environment,
its presence usually indicates that (1) the slicer module is available and
(2) the kits (mrml, vtkAddon, vtkSegmentationCore, ..) have been imported.

Error:

WARNING: autodoc: failed to import module 'parameterNodeWrapper' from module 'slicer'; the following exception was raised:
Traceback (most recent call last):
  [...]
  File "/path/to/Slicer/Base/Python/slicer/parameterNodeWrapper/__init__.py", line 5, in <module>
    from .guiConnectors import *
  File "/path/to/Slicer/Base/Python/slicer/parameterNodeWrapper/guiConnectors.py", line 16, in <module>
    from . import parameterPack as pack
  File "/path/to/Slicer/Base/Python/slicer/parameterNodeWrapper/parameterPack.py", line 10, in <module>
    from .serializers import (
  File "/path/to/Slicer/Base/Python/slicer/parameterNodeWrapper/serializers.py", line 402, in <module>
    class NodeSerializer(Serializer):
  File "/path/to/Slicer/Base/Python/slicer/parameterNodeWrapper/serializers.py", line 424, in NodeSerializer
    def write(self, parameterNode, name: str, value: slicer.vtkMRMLNode) -> None:
AttributeError: module 'slicer' has no attribute 'vtkMRMLNode'

jcfr committed:
BUG: Specify autodoc mock imports only if required

jcfr committed:
DOC: Remove irrelevant entry for api page and improve formatting

Since mrml and vtkAddon classes are associated with the slicer package,
they should be documented accordingly.

jcfr committed:
ENH: Experiment with using PythonSlicer as environment for generating doc
